diff --git a/modules/color/color.js b/modules/color/color.js
index 3814816..29473a2 100644
--- a/modules/color/color.js
+++ b/modules/color/color.js
@@ -62,6 +62,9 @@ Drupal.behaviors.color = function (context) {
     $('#text', form).css('color', inputs[4].value);
     $('#text a, #text h2', form).css('color', inputs[1].value);
 
+    // Link fill preview
+    $('#link-fill', form).css('backgroundColor', inputs[1].value);
+
     // Set up gradient
     var top = farb.unpack(inputs[2].value);
     var bottom = farb.unpack(inputs[3].value);
diff --git a/modules/color/color.module b/modules/color/color.module
index 407a93c..0865647 100644
--- a/modules/color/color.module
+++ b/modules/color/color.module
@@ -228,7 +228,7 @@ function theme_color_scheme_form($form) {
   // Preview
   $output .= drupal_render($form);
   $output .= '<h2>'. t('Preview') .'</h2>';
-  $output .= '<div id="preview"><div id="text"><h2>Lorem ipsum dolor</h2><p>Sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud <a href="#">exercitation ullamco</a> laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p></div><div id="img" style="background-image: url('. base_path() . $path . $info['preview_image'] .')"></div></div>';
+  $output .= '<div id="preview"><div id="link-fill"></div><div id="text"><h2>Lorem ipsum dolor</h2><p>Sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud <a href="#">exercitation ullamco</a> laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p></div><div id="img" style="background-image: url('. base_path() . $path . $info['preview_image'] .')"></div></div>';
 
   // Close wrapper
   $output .= '</div>';
